### ISSUE TITLE
ref.is_* - ?

### ISSUE THREAD

**Issue by kripken** [WebAssembly/binaryen#3511]
We already have `ref.is_null`, and the GC spec adds `ref.is_func, ref.is_data, ref.is_i31`. This issue is to discuss how best to implement them.

Some options:

1. Add new `RefIsFunc` etc. instructions.
2. Add a single new `RefIs` that has an immediate "what" that is an enum of `Func, Data, I31`.
3. Like 2, but add `Null` to the enum, and remove the existing `RefIsNull`.

2 and 3 add a little to the size of the instructions in memory, but that might be worth the savings in source code changes. For that reason I lean towards 3. It will also be easier to update as the spec changes.

Similiar issues will arise around `ref.as_func/data/i31` and `br_on_func/data/i31`. Note that I think some of this redundancy can be improved [in the spec](https://github.com/WebAssembly/gc/issues/169#issuecomment-765022935).

Thoughts?

**Comment by tlively** [comment]
2 and 3 seem reasonable to me. I would call the instruction class `ClassifyRef` rather than `RefIs`. `ref.is_null` is a little different from the others because it is not based on a heap type, so we would need a new `ClassifyRefOp` enum rather than just using `BasicHeapType`. I think that's reasonable, though, because not all `BasicHeapType`s correspond to real `ClassifyRef` instructions anyway, so I support 3.

**Comment by kripken** [comment]
@tlively 

Not sure I follow. I think they can all be done with one class with an enum (as in proposal 3):
```c++
class RefIs/ClassifyRef {
  enum What { Null, Func, Data, I31 } what;
  Expression* value;
};
```
That is, there is no `HeapType` or `BasicHeapType` at all. The instruction just checks if the input is null, or is a function, etc., so null looks the same as the others - ?

**Comment by tlively** [comment]
Right, I was just observing that `Func`, `Data`, and `I31` are all heap types, so we could have used `BasicHeapType` instead of a new enum to differentiate between these instructions, if not for `Null`. Whether that would have been a good idea or not is another question, and the answer doesn't really matter :)

 I think having a `ClassifyRefOp` just like we have a `BinaryOp` and `SIMDTernaryOp` is the way to go.

**Comment by kripken** [comment]
Oh, I see now, thanks.

On to bikeshedding :smile: The instructions are `ref.is_*` so it seems natural to have `RefIs` be the name of the class, to be consistent with our convention of matching the wasm name when possible. Is your concern that `RefIs` sounds ungrammatical? (which I guess it is..?) Sadly `RefTest` is already taken...

**Comment by tlively** [comment]
It's not so much that it is ungrammatical, but that `RefIs` sounds incomplete to me. Most of our instruction class names describe what kind of instructions they represent, rather than trying to reuse the literal instruction names (except when the instruction class only represents a single instruction, of course). `ClassifyRef` is a complete description of what that class of instructions does.

But I don't feel strongly about this, so if you prefer `RefIs`, I can roll with that 👍

**Comment by aheejin** [comment]
I am not very strongly opinionated but I like 2 the most, given that: `ref.is_null` is different on what it does from other three and we don't need a separate `enum` for that as @tlively said. Also I like the name `RefIs` better just because it resembles the original instruction names but I am OK with either.

**Comment by kripken** [comment]
Another option might be `RefIsX` perhaps, which is slightly more grammatical?

**Comment by kripken** [comment]
@aheejin I agree `ref.is_null` is slightly different. It feels close enough to be joined in, though, I think, and it would save a bunch of code, which is why I prefer option 3.

**Comment by tlively** [comment]
CG is now standard and the implementation is stable. Most of the instructions mentioned here did not make it into the final spec.
